File: 5. Quick Sort - GFG/5.-quick-sort.py
#User function Template for python3

class Solution:
    #Function to sort a list using quick sort algorithm.
    def quickSort(self,arr,low,high):
        # code here
        while low < high:
            p = self.partition(arr, low, high)
            self.quickSort(arr, low, p)
            low = p + 1

    
    def partition(self,arr,low,high):
        # code here
        p = arr[low]
        i = low - 1
        j = high + 1
        
        while True:
            i += 1 
            while(arr[i] < p):
                i += 1
            
            j -= 1
            while(arr[j] > p):
                j -= 1
            
            if i >= j:
                return j
            arr[i], arr[j] = arr[j], arr[i]
      
        
    

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == "__main__":
    t=int(input())
    for i in range(t):
        n=int(input())
        arr=list(map(int,input().split()))
        Solution().quickSort(arr,0,n-1)
        for i in range(n):
            print(arr[i],end=" ")
        print()

# } Driver Code Ends

File: Sorting_DSA_Basics/5. Quick Sort - GFG/5.-quick-sort.cpp
// { Driver Code Starts
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

/* Function to print an array */
void printArray(int arr[], int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

 // } Driver Code Ends
class Solution
{
    public:
    //Function to sort an array using quick sort algorithm.
    void quickSort(int arr[], int low, int high){
        // code here
        if(low < high){
            int p = partition(arr,low,high);
            quickSort(arr,low,p-1);
            quickSort(arr,p+1,high);
        
        }
    }
    
    public:
    int partition (int arr[], int low, int high){
       // Your code here
       srand(time(NULL));
       int pivot = low + rand()%(high-low);
       swap(arr[high],arr[pivot]);
       int i = low;
       for(int j = low; j < high; j++){
           if(arr[j] < arr[high]){
               swap(arr[j], arr[i++]);
           }
       }
       swap(arr[high],arr[i]);
       return i;
    }
};


// { Driver Code Starts.
int main()
{
    int arr[1000],n,T,i;
    scanf("%d",&T);
    while(T--){
    scanf("%d",&n);
    for(i=0;i<n;i++)
      scanf("%d",&arr[i]);
      Solution ob;
    ob.quickSort(arr, 0, n-1);
    printArray(arr, n);
    }
    return 0;
}  // } Driver Code Ends